Added basic template for root-password via cloud-init customization

alongoldboim opened the pull request:
Added this template for the following reasons:
* Its not trivial for user to understand that when you insert root password from the UI that you need a script in order for it to take effect, by having the basic one its clearer.
* The template implements a very basic and required cloud-init task
* Needed for the OpenShift deployment via automation, using this script to set rootpass on all vm.

Didn't understand why it has to be connected to pxe_image_type_id, if anyone has an idea please explain (I have added one that is also made in the db seed so for sure it will exist).

alongoldboim commented:
@dkorn @simon3z @gmcculloug 
Have a look please, thanks.

abonas commented on db/fixtures/customization_templates.yml:
```diff
@@ -274,3 +274,10 @@
+
+- :name: Basic root pass template
+  :description: This template takes use of rootpassword defined in the UI
+  :script: "#cloud-config\nchpasswd:\n  list: |\n    root:<%= MiqPassword.decrypt(evm[:root_password]) %>\n  expire: False"
```
it is ok from security perspective to decrypt the password here?

bdunne replied:
@abonas Can cloud-init accept an encrypted value?

alongoldboim replied:
I have tried, but it seems to get the password as plain text.

kbrock replied:
I'd try

``ruby
:script: "#cloud-config\nchpasswd:\n  list: |\n    root: \"<%= MiqPassword.decrypt(evm[:root_password]) %>\"\n  expire: False"
```

kbrock replied:
my backup suggestion is `MiqPassword.try_decrypt`. the api may have decrypted `evm[:root_password]` already

alongoldboim replied:
Didn't work, there seems to be a bug of some sort, i think the initializers aren't loading like you suggested.

abonas commented:
can you add a screenshot to which part of ui do you refer here?

gmcculloug commented:
@bdunne Please review

gmcculloug commented on db/fixtures/customization_templates.yml:
```diff
@@ -274,3 +274,10 @@
+  :description: This template takes use of rootpassword defined in the UI
+  :script: "#cloud-config\nchpasswd:\n  list: |\n    root:<%= MiqPassword.decrypt(evm[:root_password]) %>\n  expire: False"
+  :type: CustomizationTemplateCloudInit
+  :pxe_image_type_id: 1
```
You should not specify the `:pxe_image_type_id` value as the target database will likely not have a matching instance.  If the appliance is running with a region other than `REGION 0` it definitely will not match.

alongoldboim replied:
OK makes sense, how I can tackle it? (and why is the pxe_image_type_id required?)
I can do something like this instead of this db:seed
```
def add_basic_root_template
    unless CustomizationTemplate.find_by_name("Basic root pass template")
      options = { :name => "Basic root pass template",
                  :description => "This template takes use of rootpassword defined in the UI",
                  :script => "#cloud-config\nchpasswd:\n  list: |\n    root:<%= MiqPassword.decrypt(evm[:root_password]) %>\n  expire: False",
                  :type => "CustomizationTemplateCloudInit",
                  :pxe_image_type_id => PxeImageType.first.id }
      CustomizationTemplate.new(options).save
    end
  end
```
but then it will be available only for users who used deployment at least once.

bdunne replied:
@alongoldboim I believe you can just remove that line in the YAML.

gmcculloug replied:
Correct, following the pattern of what is already there you want to set `:system` to true and when it is seeded it will be available in the `Examples` area of the Customization Templates.

bdunne commented on db/fixtures/customization_templates.yml:
```diff
@@ -274,3 +274,10 @@
+  :script: "#cloud-config\nchpasswd:\n  list: |\n    root:<%= MiqPassword.decrypt(evm[:root_password]) %>\n  expire: False"
+  :type: CustomizationTemplateCloudInit
+  :pxe_image_type_id: 1
+  :system: nil
```
`:system:` should be true, that will prevent users from editing it.

alongoldboim commented:
@bdunne @gmcculloug 
* changed system to true
* removed pxe_image_type_id

seems to work well now.
@abonas, added a default template.
![selection_002](https://cloud.githubusercontent.com/assets/11405684/13746392/f4962ac8-e9fb-11e5-86c8-cf0b5b23705d.png)

bdunne commented:
:+1: LGTM.  Thanks @alongoldboim

kbrock commented:
@gmcculloug can you confirm that rails initializers are not loaded in automate tasks (intentionally)

gmcculloug commented:
@kbrock The issue @alongoldboim is running into is likely happening during substitution of the customization template which is part of the backend provisioning state-machine, not automate.  Automate is used to submit the request using `create_provision_request` and the `:root_password` is provided as a string which automate is not processing.  We can easily validate the root_password makes it into the miq_request object correctly.  

Also, the error `[MiqPassword::MiqPasswordError]: can not decrypt v0_key encrypted string` is pointing to a string that does not have the `v1_` or `v2_` prefix which means it may already be decrypted.

@alongoldboim Do you have a stack trace when you get the error?

alongoldboim commented:
@gmcculloug If you need more help besides the session we had please ping me.

gmcculloug commented:
Issue #7285 is unrelated to this PR since it does work when submitted from the UI.  The issue can be worked separately.

@Fryguy Any concerns from the security side here since the script is using `MiqPassword.decrypt(evm[:root_password])` which will use resolve the plain-text password into the cloud-init script?

bdunne commented:
:+1: LGTM  I wouldn't call that a security risk since it's an example script and I think cloud-init requires a clear-text password.

Fryguy commented:
No, it's not a security risk...it's how cloud-init works.
